ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -224,26 +230,48 @@ public PullQueryResult execute(
           .getMaterialization(queryId, contextStacker)
           .orElseThrow(() -> notMaterializedException(getSourceName(analysis)));
 
-      final Struct key = asKeyStruct(whereInfo.keyBound, query.getPhysicalSchema());
+      final List<Optional<KsqlNode>> sourceNodes = new ArrayList<>();
+      final List<List<?>> tableRows = new ArrayList<>();
+      final List<LogicalSchema> schemas = new ArrayList<>();
+      final List<Future<PullQueryResult>> futures = new ArrayList<>();
+      final List<List<Struct>> keysByLocation = mat.locator().groupByLocation(
+          whereInfo.keysBound.stream()
+              .map(keyBound -> asKeyStruct(keyBound, query.getPhysicalSchema()))
+              .collect(Collectors.toList()));
+
+      for (List<Struct> keys : keysByLocation) {
+        final PullQueryContext pullQueryContext = new PullQueryContext(
+            keys,
+            mat,
+            analysis,
+            whereInfo,
+            queryId,
+            contextStacker,
+            pullQueryMetrics);
+
+        futures.add(executorService.submit(() -> handlePullQuery(
+            statement,
+            executionContext,
+            serviceContext,
+            pullQueryContext,
+            routingOptions
+        )));
 
-      final PullQueryContext pullQueryContext = new PullQueryContext(
-          key,
-          mat,
-          analysis,
-          whereInfo,
-          queryId,
-          contextStacker,
-          pullQueryMetrics);
+      }
+      for (Future<PullQueryResult> future : futures) {
+        final PullQueryResult result = future.get();
```

Comment:
Shouldn't there be some error handling here? What happens if a thread dies due to an uncaught exception? We would like to fail the entire query and not return partial results, right? Maybe handle something like this:
```
        try {
            future.get();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } catch (ExecutionException e) {
            // Extract the actual exception from its wrapper
            Throwable t = e.getCause();
            System.err.println("Uncaught exception is detected! " + t
                    + " st: " + Arrays.toString(t.getStackTrace()));
            // ... Handle the exception
        }
```

Reply:
Yeah, I agree that we don't want partial results, so I wouldn't want to catch each individual future.

Also, calling interrupt is what causes the `InterruptedException` and it sends that exception to other threads.  To have the current thread have it thrown, all we have to do is not catch it.

I pretty much just let the existing handler catch Exception in the wider scope.  I added a small special case to unwrap `ExecutionException` so we get the same error message.

Reply:
> calling interrupt is what causes the `InterruptedException`

I think you probably meant this, but just a small clarification here - `Thread.currentThread().interrupt()` does not directly cause an `InterruptedException`. It just sets an interrupt flag on the thread's state, and then any blocking code is expected to check that flag and raise an `InterruptedException` when it notices that flag. There is no guarantee that calling `interupt` will actually do anything (especially if there's misbehaving client code). It's a cooperative strategy.

Reply:
Yeah, you're right that a thread may have to explicitly check for being interrupted if it's doing a lot of computation.  If it's sleeping in some manner, there's a good chance that it will be thrown for the thread, e.g. `Thread.sleep`.